### PR TITLE
Added tracking when a class is constructed (new).

### PR DESCRIPTION
**Pull request by terrylucas**
Sometimes it is difficult to know which code is leaking or causing too much memory to be allocated. Allocation monitoring was created to see the delta of objects between two points in time.  However, even this useful tool can be hard to pinpoint who or what is causing the memory explosion.

To monitor a particular allocation first start the allocation monitoring by pressing

image

Using the Search field, type in the name of a class to track e.g., ObjectWithUniqueId

image

Adding one of your known classes in your application being profiled in this example "ObjectWithUniqueId" is typed in

image

Selecting the class to the tracking will scroll the class into view and enable the checkbox will add the class to the list of classes to track:

image

now interact with the application your profiling to track the class(es) that might be constructed.

now press the monitor button image again

At the bottom on the monitor pane (right side) are the track classes, expanding the tracked class will display all instances of  classes that were constructed (between when the class to added to tracking list and the monitor button was pressed). In the below example, 14 instances where constructed.

image

Expanding on when (timestamp) the construction occurred will display the call stack of the top 4 entries when the constructor was called:

image

Clicking on "More ..." will display the entire call stack.

image

### CODE REVIEW

**Comment by jacob314** [comment]
High level UX feedback:
I would expect most of the time I want to track a class after noticing something troubling about the objects allocated of the class.
Perhaps add a column of "track" checkboxes similar to the checkbox icon you currently show?
Do you have an example problem app with a problem that can be solved with the tracker data? That could help us iterate on the UI with some concrete examples in mind. Having it be a separate table makes me nervous particularly if a user is tracking allocations for multiple different classes but wants to drill down into allocations for just one of them.
I would suggest we instead integrate it with the existing allocation tracker display now and in the future with the snapshot display once we can associate allocations with snapshots. For example, from the allocation tracker table I would want to see all allocations for the class I selected.
The stack trace for an allocation means nothing to me unless I know what class was allocated. Maybe I'm missing where that is in the screenshots.

File: case_study/memory_leaks/images_1/lib/main.dart
Comment (jacob314) [inline case_study/memory_leaks/images_1/lib/main.dart:41]: nit: can we use a different name as this sounds like the name of the feature it is testing which makes the UI confusing. Maybe just something simple like
ObjectWithUniqueId.
I think my confusion in previous comments was from seeing the word Tracker in the table.

File: packages/devtools_app/lib/src/memory/memory_allocation_table_data.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_allocation_table_data.dart:12]: lets use a real checkbox and make these clickable.

File: packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart:73]: why does this fire if the tree hasn't changed?

File: packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart:80]: This should use ListView or preferably our TreeTable class instead. As written this will kill Flutter Web performance once there are a lot of allocations.
If we need to add some features to the TreeTable I think that would be world be worth it. For example, having an option to truncate to only show the first N (4) children by default and support expanding to show the rest of the content.

File: packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart:175]: use the existing size constants. there is a normalSpacing that is 16 and small that is 8. Using those will keep the UI a little more consistent.

File: packages/devtools_app/lib/src/memory/memory_controller.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_controller.dart:111]: var -> bool

File: packages/devtools_app/lib/src/memory/memory_controller.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_controller.dart:125]: Fyi @kenzieschmoll I think you have wanted to filter stack traces from the Dart core libraries and Flutter as well. @terrylucas is this method used currently? I suspect implementing it will be hard as the stacktraces currently lack the info needed.
One other item to hide and show is native frames.
This is something we should filter consistently across DevTools.

File: packages/devtools_app/lib/src/memory/memory_controller.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_controller.dart:252]: maybe add a comment on what this does?

File: packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart:667]: Is there a use case where we need this instead of manually selecting classes?

File: packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart:706]: we need to display a toast at least the first time this is done that the results only show up for new allocations.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:238]: Fyi @bkonyi we need to make these or similar APIs public instead of using a private API.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:256]: @bkonyi same issue with this API

**Comment by terrylucas** [comment]
> High level UX feedback:
> I would expect most of the time I want to track a class after noticing something troubling about the objects allocated of the class.
> Perhaps add a column of "track" checkboxes similar to the checkbox icon you currently show?
> Do you have an example problem app with a problem that can be solved with the tracker data? That could help us iterate on the UI with some concrete examples in mind. Having it be a separate table makes me nervous particularly if a user is tracking allocations for multiple different classes but wants to drill down into allocations for just one of them.
> I would suggest we instead integrate it with the existing allocation tracker display now and in the future with the snapshot display once we can associate allocations with snapshots. For example, from the allocation tracker table I would want to see all allocations for the class I selected.
> The stack trace for an allocation means nothing to me unless I know what class was allocated. Maybe I'm missing where that is in the screenshots.

Checkboxes looked like a bit of work to add in TableRow, I do think it's a better UI.  However the scenario you're suggesting is how it works.  I think the confusion might be that I added a Tracker to the case study to test where find when the user's class e.g., Tracker as found.

File: case_study/memory_leaks/images_1/lib/main.dart
Comment (terrylucas) [reply case_study/memory_leaks/images_1/lib/main.dart:41]: Done.

**Comment by jacob314** [comment]
> > High level UX feedback:
> > I would expect most of the time I want to track a class after noticing something troubling about the objects allocated of the class.
> > Perhaps add a column of "track" checkboxes similar to the checkbox icon you currently show?
> > Do you have an example problem app with a problem that can be solved with the tracker data? That could help us iterate on the UI with some concrete examples in mind. Having it be a separate table makes me nervous particularly if a user is tracking allocations for multiple different classes but wants to drill down into allocations for just one of them.
> > I would suggest we instead integrate it with the existing allocation tracker display now and in the future with the snapshot display once we can associate allocations with snapshots. For example, from the allocation tracker table I would want to see all allocations for the class I selected.
> > The stack trace for an allocation means nothing to me unless I know what class was allocated. Maybe I'm missing where that is in the screenshots.
> 
> Checkboxes looked like a bit of work to add in TableRow, I do think it's a better UI. However the scenario you're suggesting is how it works. I think the confusion might be that I added a Tracker to the case study to test where to find the user's class e.g., Tracker as found.

You should be able to add checkboxes to a table row by specifying a builder for the ui for each entry in the row. If that is hard for the current table impl, we should clean it up. The table used in the logging view shows some examples of this.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (bkonyi) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:216]: Why is this being changed to `_getAllocationProfile`? The current string references the existing public RPC.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (bkonyi) [reply packages/devtools_app/lib/src/vm_service_wrapper.dart:249]: Yeah, I'd strongly prefer that we avoid using private methods outside of VM developer mode. I'm working on a public implementation now, so it would be great if we can wait on landing this PR until we have that in place. I'm much more comfortable having a DevTools out in the wild that checks to see if there's support for `setTraceClassAllocation` and falls back to `_setTraceClassAllocation` than just blindly calling `_setTraceClassAllocation`, which will likely be renamed and break with newer service targets.

**Review by bkonyi — CHANGES_REQUESTED** [review]
I'm working on making `_setTraceClassAllocation` public, so please wait until support has landed in `package:vm_service` before landing.

I'll be checking this in.  We wont we publishing devtools for a number of weeks so once you checkin new vm_service I'll update to use published name.  Will not have version.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (bkonyi) [reply packages/devtools_app/lib/src/vm_service_wrapper.dart:267]: I also have a public implementation in the works for this.

File: packages/devtools_app/lib/src/memory/memory_allocation_table_data.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_allocation_table_data.dart]: Done.

File: packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart]: Yeah I've re-written using TreeTable can't nest treeTable in another treetable but I'll look at how to change Table to support that.

File: packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart]: Re-wrote using TreeTable not needed.

File: packages/devtools_app/lib/src/memory/memory_controller.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_controller.dart:111]: Done.

File: packages/devtools_app/lib/src/memory/memory_controller.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_controller.dart]: Done.

File: packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart]: Re-wrote and eliminated using TreeTable and auto-complete search.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/vm_service_wrapper.dart:256]: okay, let me know when its ready.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/vm_service_wrapper.dart:274]: okay, ditto let me know when ready.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/vm_service_wrapper.dart]: It was accidentally changed - shouldn't have been.

File: packages/devtools_app/lib/src/memory/memory_tracker_model.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_tracker_model.dart:200]: this looks like test code move it to a test directory file.

File: packages/devtools_app/lib/src/memory/memory_tracker_model.dart
Comment (jacob314) [inline packages/devtools_app/lib/src/memory/memory_tracker_model.dart:86]: this seems odd. why does the children of a contain a? Add a comment explaining more what is going on.
Naively seems like we are just returning -1 a lot.

**Review by jacob314 — APPROVED** [review]
lgtm

File: packages/devtools_app/lib/src/memory/memory_tracker_model.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_tracker_model.dart:86]: Yes it is odd in many ways.  Returning -1 happens to not sort (which is what I wanted).  Also,  `valueB = b.children.indexOf(a)` is wrong too.  Must have been a brain fart.

The two lines should have been:

```
      valueA = a.parent.children.indexOf(a);
      valueB = b.parent.children.indexOf(b);
```

File: packages/devtools_app/lib/src/memory/memory_tracker_model.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_tracker_model.dart:200]: Done.

File: packages/devtools_app/lib/src/memory/memory_controller.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_controller.dart:125]: This is currently controlled by the filter dialog in memory profiler.  Should this filtering be a a higher level (top-level setting)?  Then all stack traces, classes displayed in snapshots, allocation monitoring, etc. would use the same filtering.

I'll add an issue https://github.com/flutter/devtools/issues/2685

File: packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_allocation_table_view.dart:81]: It only fires if the tree has changed true or false.  However, the tree never resets back to false once fired (value == true). So the test is meaningless - removed.

File: packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart
Comment (terrylucas) [reply packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart]: Added TODO.  We have a message bus for toast but I don't see a toast to display when event is received. Do we have a toast in DevTools (I thought we did maybe it never got ported from HTML to Flutter).

**Comment by bkonyi** [comment]
Can we make sure this isn't published in any release until [this CL](https://dart-review.googlesource.com/c/sdk/+/182666) lands? This change as-is will break with newer VM service implementations once it lands.

**Comment by terrylucas** [comment]
Yes, this will not be published for a few weeks (2/24).  We should have a
new VMService by then - right?

**Comment by bkonyi** [comment]
Yes, we should. I'm still waiting on a review for the CL above but I think Ryan is occupied with something else at the moment. I'll push harder for a review if it's not ready today.

**Pull request merged**
